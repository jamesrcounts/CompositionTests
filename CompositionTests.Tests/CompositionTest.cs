using System;
using System.ComponentModel.Composition;
using System.ComponentModel.Composition.Hosting;
using System.IO;
using System.Reflection;
using Microsoft.ComponentModel.Composition.Diagnostics;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace CompositionTests.Tests
{
    [TestClass]
    public class CompositionTest
    {
        [TestMethod]
        public void TestDiscoverParts()
        {
            Composition.DiscoverParts(
                new AssemblyCatalog(Assembly.GetExecutingAssembly()),
                s => s.ScrubVersionNumber().ScrubPublicKeyToken());
        }

        [TestMethod]
        public void TestDiscoverPartsWithCatalogAndProvider()
        {
            AssemblyCatalog catalog = new AssemblyCatalog(Assembly.GetExecutingAssembly());
            CompositionContainer compositionContainer = new CompositionContainer(catalog);
            Composition.DiscoverParts(
                catalog,
                compositionContainer,
                s => s.ScrubPublicKeyToken(),
                s => s.ScrubVersionNumber());
        }

        [TestMethod]
        public void TestDiscoverPartsWithFactory()
        {
            Composition.DiscoverParts(
                () =>
                {
                    var catalog = new AssemblyCatalog(Assembly.GetExecutingAssembly());
                    return new CompositionInfo(catalog, new CompositionContainer(catalog));
                },
                s => s.ScrubPublicKeyToken(),
                s => s.ScrubVersionNumber());
        }

        [TestMethod]
        public void TestReflectionTypeLoadException()
        {
            var consoleError = Console.Error;
            var errorTrap = new StringWriter();
            try
            {
                Console.SetError(errorTrap);
                Composition.DiscoverParts(() =>
                {
                    string message = "This is an inner exception.";
                    throw new ReflectionTypeLoadException(new[] { this.GetType() }, new[] { new InvalidOperationException(message) });
                });
                Assert.Fail("Expected Exception");
            }
            catch (ReflectionTypeLoadException)
            {
                ApprovalTests.Approvals.Verify(errorTrap.ToString());
            }
            finally
            {
                Console.SetError(consoleError);
            }
        }

        [Export]
        public class TestingPart
        {
            [Import]
            public int MyProperty { get; set; }
        }
    }
}