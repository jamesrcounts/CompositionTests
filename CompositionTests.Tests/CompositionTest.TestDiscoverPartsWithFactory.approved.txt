[Part] CompositionTests.Tests.CompositionTest+TestingPart from: AssemblyCatalog (Assembly="CompositionTests.Tests, Culture=neutral")
  [Primary Rejection]
  [Export] CompositionTests.Tests.CompositionTest+TestingPart (ContractName="CompositionTests.Tests.CompositionTest+TestingPart")
  [Import] CompositionTests.Tests.CompositionTest+TestingPart.MyProperty (ContractName="System.Int32")
    [Exception] System.ComponentModel.Composition.ImportCardinalityMismatchException: No exports were found that match the constraint: 
    ContractName    System.Int32
    RequiredTypeIdentity    System.Int32
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition)
   at Microsoft.ComponentModel.Composition.Diagnostics.CompositionInfo.AnalyzeImportDefinition(ExportProvider host, IEnumerable`1 availableParts, ImportDefinition id)

[Part] CompositionTests.Tests.FileRemover from: AssemblyCatalog (Assembly="CompositionTests.Tests, Culture=neutral")
  [Export] CompositionTests.Tests.FileRemover (ContractName="CompositionTests.Tests.IRemoveFiles")

